### PR TITLE
Add TechdocsContainerRunner interface as successor of deprecated ContainerRunner for TechdocsGenerator

### PR DESCRIPTION
**Pull request by JoergSiebahn**
## Hey, I just made a Pull Request!

Based on the discussion in #24928 the `ContainerRunner` has [been deprecated](https://github.com/backstage/backstage/commit/f2f41aad7e6152f13c9b5d50674511382847f298). 

In the current state, once the `ContainerRunner` interface is removed, the only alternative to generate TechDocs in another environment (e.g. Kubernetes or any kind of workflow engine) would be to replace the `TechdocsGenerator` with a complete individual implementation. In that case one has also to take care about `mkdocs.yml` validation and modification as well as the post processing. Both seem crucial for security and the TechDocs cache and workflow and should imho not be part of individual implementations when the only goal is another execution environment.

This PR aims to provide the `ContainerRunner` interface specifically for TechDocs generation in `plugin-techdocs-node` as `TechdocsContainerRunner`, so that `ContainerRunner` in `backend-common` can be removed one day and `TechdocsGenerator` still supports individual execution environments.

The new interface and types still support the original `ContainerRunner` interface so that this change is not breaking. I would appreciate that this can be confirmed by an expert, as I'm mostly a Java developer and quite new to TypeScript.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by JoergSiebahn** [comment]
Imho this it not stale unless there is a decision to drop support for custom container environments. @Rugvip, @camilaibs, @vinzscam could one of you have a look? Just pinging you, because you were already part of the discussion and I don't know who could be responsible for .

File: plugins/techdocs-node/src/stages/generate/types.ts
Comment (camilaibs) [inline plugins/techdocs-node/src/stages/generate/types.ts:63]: This types should not extends the `ContainerRunner` since it is deprecated and will be deleted. My recommendation is to copy the `ContainerRunner` props you need to the `TechdocsContainerRunner` you created.

**Review by camilaibs — COMMENTED** [review]
Thanks for contributing with us and sorry for the delay in reviewing your pull request, we are just back from summer vacations... Just left a minor comment 🙂

File: plugins/techdocs-node/src/stages/generate/types.ts
Comment (JoergSiebahn) [reply plugins/techdocs-node/src/stages/generate/types.ts]: Hi @camilaibs,

it's summer time. I'm also just back from vacation.

Thanks for your review. I was trying to maintain backward compatibility until the `ContainerRunner` is finally removed. But you are right, that doesn't make sense here. Both interfaces share nearly the same declaration. Only `RunContainerOptions` has an optional `pullOptions` property that is not used any more and does not exist in the new `TechdocsContainerRunOptions` as in [the implementation of `DockerContainerRunner`](https://github.com/backstage/backstage/pull/25543/files#diff-b131b4ff78b97d217515ac30ebaaa3d2323e893a076189dbd5fced9c4515ed2cL41) before. I think there are more places where imports from backend-common can be removed or replaced by the Techdocs* types and interfaces introduced here.

- the [`TechdocsRunContainerOptions`](https://github.com/backstage/backstage/pull/25543/files#diff-e9ac315c875593eddfd73665103aeefb394528a2a6a2564870b808750cd323f8R331) may not need to allow `RunContainerOptions` as in the previous state of this PR
- the [static `fromConfig`](https://github.com/backstage/backstage/blob/master/plugins/techdocs-node/src/stages/generate/generators.ts#L45) method of the generator takes a `ContainerRunner` but may also take a `TechdocsContainerRunner`

I updated the PR accordingly. May I get another review?

**Comment by JoergSiebahn** [comment]
Hi @camilaibs,

while rebasing this PR I realized that you already covered most of the proposed changes in #26080.

This PR dropped down to the following:

- Move interface `TechDocsContainerRunner` from publish to generate package. I think, it makes more sense there, because the container runner is used by generate, not by publish.
- A little bit of documentation.
- Explicit implementation of `TechDocsContainerRunner` by `DockerContainerRunner`. Is there a reason to skip the `implements` declaration in TypeScript? (remember my Java background 😉)

Do you think it's worth to merge these little changes?

**Review by camilaibs — APPROVED** [review]
Thanks so much for this refactoring, it looks great 💯

**Comment by camilaibs** [comment]
PTAL @backstage/techdocs-maintainers

**Review by byan1197 — APPROVED** [review]

**Pull request merged**
